### PR TITLE
Emacs mode lake

### PR DESCRIPTION
**Pull request by lecopivo**
A simple command for running `lake build` in lean4-mode

### CODE REVIEW

**Comment by Kha** [comment]
Thanks. Could you add the keyboard shortcut to `lean4-mode/README.md`?

File: lean4-mode/lean4-lake.el [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+(defun lean4-lake-find-dir-safe ()
+  (or (lean4-lake-find-dir)
+      (error (format "cannot find lakefile.lean for %s" (buffer-file-name)))))
+
+(defun lean4-lake-build ()
+  "Call lake build"
+  (interactive)
+  (let* ((default-directory (file-name-as-directory (lean4-lake-find-dir-safe))))
+    (with-existing-directory
+      (compile "lake build"))))
```
Comment (Kha) [inline lean4-mode/lean4-lake.el:19]: You should probably use `(lean4-get-executable "lake")` here

**Comment by lecopivo** [comment]
Sure, I fix that. I will also add ability to run `lake run <script>`. Would something else make sense?

**Comment by tydeu** [comment]
@lecopivo

> I will also add ability to run `lake run <script>`.

Fyi, I plan to change this to `lake script run <script>` soon, so maybe hold off on that for just a little while?

**Comment by lecopivo** [comment]
Sure, I can wait. 

Also, is there a planned support for unit testing in lake? Or is that ment to be done through scripts?

**Comment by tydeu** [comment]
@lecopivo 

> Also, is there a planned support for unit testing in lake? Or is that ment to be done through scripts?

Yeah, testing will probably be done through a combination of a separate testing library, a test runner executable, and a script (like how [mocha](https://mochajs.org/) is generally used in JavaScript).

**Comment by Kha** [comment]
You should use `lake env` (if there is a lakefile) for running `lean` with `LEAN_PATH` set up correctly.

**Comment by lecopivo** [comment]
> You should use `lake env` (if there is a lakefile) for running `lean` with `LEAN_PATH` set up correctly.

Ohh I didn't about this, well at least I got some practice with elisp after a long time.

**Comment by lecopivo** [comment]
Concerning keybindings, I'm not the biggest fan of the `C-c C-p C-l`. I picked it to be consistent with keybindings for leanpkg. What about having `C-c C-x` to be `lean4-std-exe` and `C-c C-l` be `lean4-lake-build`. Alternatively, we can use `C-c C-b` for `lean4-lake-build` as `lean4-message-boxes-toggle` does no longer exist.

**Comment by Kha** [comment]
Either would be fine with me.

**Comment by lecopivo** [comment]
I will go for `C-c C-l` for `lean4-lake-build`. I will add an option to rebuild lake project before running `lean4-std-exe`.

**Comment by Kha** [comment]
I don't have a strong opinion on any of these since I never use them :) . But rebuilding might be problematic if some downstream module fails to build. Ideally we would only rebuild the imports like the server does via `lake print-paths`, but that's a bit more complicated.

**Review by Kha — CHANGES_REQUESTED** [review]
> You should use `lake env` (if there is a lakefile) for running `lean` with `LEAN_PATH` set up correctly.

This is still missing I think

**Comment by lecopivo** [comment]
Yes it is still missing. I will try to fix that and change the keybindings sometime this week. In the current form it solves my problem :) so I had priorities elsewhere.

**Pull request closed without merging**